Add PMIX_QUERY_ABI_VERSION (#397)
pmix/pmix-standard: Chap_API_Sync_Access.tex

The returned \refarg{status} indicates if requested data was found or not. The returned \refarg{info} array will contain a \refattr{PMIX_QUERY_RESULTS} element for each query of the \refarg{queries} array. If qualifiers were included in the query, then the first element of each results array shall contain the \refattr{PMIX_QUERY_QUALIFIERS} key with a \refstruct{pmix_data_array_t} containing the qualifiers. The remaining \refstruct{pmix_info_t} shall contain the results of the query, one entry for each key that was found. Note that duplicate keys in the \refarg{queries} array shall result in duplicate responses within the constraints of the accompanying qualifiers. The caller is responsible for releasing the returned array.

> The \refapi{PMIx_Query_info} and \refapi{PMIx_Query_info_nb} operations are expected to be called between initialization (e.g., \refapi{PMIx_Init}) and finalization (e.g., \refapi{PMIx_Finalize}) with the exeption of calling with exclusively with the following set of query attributes:

Author: ```suggestion
The \refapi{PMIx_Query_info} operation is expected to be called between initialization (e.g., \refapi{PMIx_Init}) and finalization (e.g., \refapi{PMIx_Finalize}) with the exeption of calling the blocking query \ac{API} exclusively with the following set of query attributes:
```

\declareAttributeNEW{PMIX_SERVER_INFO_ARRAY}{"pmix.srv.arr"}{pmix_data_array_t}{
Array of \refstruct{pmix_info_t} about a given server, starting with its \refattr{PMIX_NSPACE} and including at least one of the rendezvous-required pieces of information.
}
%
\declareAttributeNEW{PMIX_QUERY_ABI_VERSION}{"pmix.qry.abiversion"}{char *}{
Query the \ac{ABI} version supported by the \ac{PMIx} library. The version returned will be of the form \code{"MAJOR.MINOR"}. For example. \code{"4.2"} representing version v4.2 of the \ac{PMIx} \ac{ABI}. NO QUALIFIERS.

> }

Author: Notes from April 14, 2022 meeting
 * Good discussion about how the "Major.Minor" will be updated and how an application might use this information.
 * What is the guidance on updating the "Major.Minor"?
   * If a new API is added, this does not break the ABI but does change the ABI. So increment the "Minor".
   * If an API is removed, this does break the ABI. So increment the "Major" and set "Minor" to 0.
   * If an API is modified, this does break the ABI. So increment the "Major" and set "Minor" to 0.
 * If an application builds against "`1.7`" of the ABI.
   * They require a major version of "`1`" and a minor version of "`x >= 7`" since they do not care about additions to the ABI only that the `1.7` ABI is still intact.
   * If the major version is not "`1`" then they need to consult the ABi to determine if the modifications/removals impact their application.
 * The complexity comes with Provisional items that are allowed to change form between minor revisions. This is in contrast to Stable items that have a longer Deprecation/Removal process.
   * If a Provisional API is changed then it changes the "Major" number in the ABI version causing broad disruption.
   * The suggestion was to separate the Stable and Provisional interfaces (as best as possible) and have two queries. `PMIX_QUERY_ABI_VERSION_STABLE` and `PMIX_QUERY_ABI_VERSION_PROVISIONAL`. This will help isolate applications focused on stable interfaces from rapid development of provisional items.
 * In some circumstances an implementation may continue to provide deprecated/removed interfaces. In this case they support multiple versions of the ABI. Consider two ABI versions `1.7` and `2.0` where the only difference is the removal of `PMIx_Foo()`
   * They support the old version `1.7` since it still includes `PMIx_Foo()`.
   * They also support the new version `2.0`. They just have an additional function `PMIx_Foo()` that is not defined in the `2.0` ABI.
   * As such it is important to allow an implementation to report more than one ABI level.
   * Suggestion: allow for a comma separated list to be returned.
   * If the difference between `1.7` and `2.0` is a modification of an existing API/structure then an implementation must determine which ABI they conform to. The implementation might choose to add a configure time option to choose the ABI level. Other runtime level options are available though more complex to implement.

Author: FYI. I added this to the Governance language
 * https://github.com/pmix/governance/pull/34/files#r851349685
I'm going to suggest that we also add it to the PMIx Standard (in PR #375) then reference that section here

Author: `pmix-abi` ticket to identify the Stable vs Provisional ABI:
 * https://github.com/pmix/pmix-abi/issues/11

Author: I pushed a new commit to address this comment.
